orttraining/orttraining/training_ops/cuda/optimizer/adamw/adamw_impl.cu

```diff
@@ -0,0 +1,222 @@
+
+    // Compute the new weight.
+    const float denom = (_Sqrt(m2) / _Sqrt(beta_correction)) + epsilon;
+    w = w - (lr * m1) / (alpha_correction * denom);
+
+    // Update the new weight and momentums.
+    weight_chunk_ptr[i] = static_cast<T_WEIGHT>(w);
+    momentum_1_chunk_ptr[i] = static_cast<T_MOMENTUM>(m1);
+    momentum_2_chunk_ptr[i] = static_cast<T_MOMENTUM>(m2);
+  }
+}
+
+// Huggingface AdamW equivalence.
+template <typename T_WEIGHT, typename T_GRAD, typename T_MOMENTUM>
+__global__ void AdamWComputeMode1(
+    ChunkGroup<MTA_ADAMW_GROUP_SIZE> chunks,
+    const float alpha,
+    const float beta,
+    const float epsilon,
+    const float lr,
+    const float lr_corrected,
+    const float decay) {
+  const int block_idx = blockIdx.x;
+
+  T_WEIGHT* weight_chunk_ptr;
+  T_GRAD* grad_chunk_ptr;
+  T_MOMENTUM* momentum_1_chunk_ptr;
+  T_MOMENTUM* momentum_2_chunk_ptr;
+  int chunk_size;
+
+  PrepareMTAData(chunks, block_idx, weight_chunk_ptr, grad_chunk_ptr,
+                 momentum_1_chunk_ptr, momentum_2_chunk_ptr, chunk_size);
+
+#pragma unroll(4)
+  for (int i = threadIdx.x; i < chunk_size; i += blockDim.x) {
+    float w = static_cast<float>(weight_chunk_ptr[i]);
+    float g = static_cast<float>(grad_chunk_ptr[i]);
+    float m1 = static_cast<float>(momentum_1_chunk_ptr[i]);
+    float m2 = static_cast<float>(momentum_2_chunk_ptr[i]);
+
+    // Compute exponentially-averaged historical gradient.
+    m1 = alpha * m1 + (1.f - alpha) * g;
+
+    // Compute exponentially-averaged historical squared gradient.
+    m2 = beta * m2 + (1.f - beta) * g * g;
+
+    float denom = _Sqrt(m2) + epsilon;
+    w = w - (lr_corrected * m1 / denom);
+
+    // Perform weight decay.
+    w = w - (lr * decay * w);
+
+    // Update the new weight and momentums.
+    weight_chunk_ptr[i] = static_cast<T_WEIGHT>(w);
+    momentum_1_chunk_ptr[i] = static_cast<T_MOMENTUM>(m1);
+    momentum_2_chunk_ptr[i] = static_cast<T_MOMENTUM>(m2);
+  }
+}
+
+template <typename T_WEIGHT, typename T_GRAD, typename T_MOMENTUM>
+void AdamWMTAFunctor<T_WEIGHT, T_GRAD, T_MOMENTUM>::operator()(
+    cudaStream_t stream,
+    ChunkGroup<MTA_ADAMW_GROUP_SIZE> chunks,
+    const float alpha,
+    const float beta,
+    const float epsilon,
+    const float lr,
+    const float decay,
+    const int64_t adam_mode,
+    const int64_t correct_bias,
+    const int64_t update_count) {
+  const int block_count = chunks.chunk_count;
+  const int thread_count = ChunkGroup<MTA_ADAMW_GROUP_SIZE>::thread_count_per_block;
+
+  float alpha_correction = 1.f, beta_correction = 1.f;
+  float lr_corrected = lr;
+  if (correct_bias == 1) {
+    // Notes:
+    // > there is a minor difference compared with Apex's implementation,
+    //   which uses double storing corrections before casting to float passing to kernels.
+    // > std::pow(float, int) return double since C++11, so we cast back to float.
+    alpha_correction = 1.f - static_cast<float>(std::pow(alpha, update_count));
```

Comment:
will we get better precision by casting to float after the subtraction? 
std::pow(alpha, update_count) is a small number << 1, so  a precision loss will affect it much more than (1-<a very small number)>) will

Reply:
Yeah, it might be. While as the comment says, it is to match what we see from other frameworks. 
I original use double to calculate them, then changed this from double to float, to avoid for some rare case, that might bring differences between our runs and torch/HF runs. Make sense?

Reply:
nit: The comment is not clear if we are matching apex or diverging from it, so if you have other things to fix, you can make the comment a bit clearer too.